library/system/io.lean

```diff
@@ -68,6 +91,14 @@ interface.iterate e α a f
 def forever {e} (a : io_core e unit) : io_core e unit :=
 iterate () $ λ _, a >> return (some ())
 
+def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
+interface.catch e₁ e₂ α a b
+
+instance : alternative io :=
+{ interface.monad _ with
+  orelse := λ _ a b, catch a (λ _, b),
+  failure := λ _, io.fail "failure" }
+
```

Comment:
> I think we should add a new type class [`exceptional`] that has the catch and fail actions

If we do that, then `alternative` should extend `exceptional`.  For me the main value of the alternative type-class is that we can avoid duplicate definitions, e.g. `optional`, or other abstractions that work in all of parser/parser/tactic/io/solver/smt_tactic/etc.

Having said that, at the moment alternative extends applicative and not monad, I think we need to change that since we can't get the laws for bind from the laws for seq.  AFAICT these are the possible laws:
 1. `<|>` is associative
 2. `(failure <|> a) = a`
 3. `(a <|> failure) = a`
 4. `(a >> failure) = failure`
 5. `(failure >>= a) = failure`
 6. `>>` is left-distributive over `<|>`
 7. `>>=` is right-distributive over `<|>`
 8. `(return x <|> a) = return x`

The monads `list` and `set` satisfy all of these laws.   The `option` and `result` monads satisfies all except 7.

On the other hand, `io`, `parser`, `tactic` should still (or could be made to) satisfy all laws except 4, 6, and 7.

The restriction to monad doesn't seem so bad.  We could have the following:
```lean
class exceptional (f : Type u → Type v) extends monad f :=
(failure : Π {α : Type u}, f α)
(orelse  : Π {α : Type u}, f α → f α → f α)
-- laws 1-3, 5, 8

class alternative (f : Type u → Type v) extends exceptional f :=
-- laws 4,6,7
```

I'm not sure if we'd ever use the additional laws provided by alternative, or what we'd gain at all from adding the laws.